Stable Releases: Definition and Process (#6394)
cosmos/cosmos-sdk: CONTRIBUTING.md

    - [Development Procedure](#development-procedure)
    - [Pull Merge Procedure](#pull-merge-procedure)
    - [Release Procedure](#release-procedure)
    - [Stable Release Update Procedure](#stable-release-update-procedure)

>     - [Stable Release Updates](#stable-release-updates)

Reviewer: This seems inconsistent with the recent updates calling it "point release" again.

@alessio can you update this file accordingly?

Author: Fixed. Will double check URLs sanity shortly.

>     - [Stable Release Update Procedure](#stable-release-update-procedure)
>     - [Stable Release Updates](#stable-release-updates)

Author: ```suggestion
    - [Point Release Procedure](#point-release-procedure)
```

- Tag the release (use `git tag -a`) and create a release in Github
- Delete the `RC` branches

> ### Point Release Procedure

Reviewer 2: I would really prefer to keep the nomenclature point releases because that is what they are. In fact, in very rare cases, they include breaking changes.

Author: Introducing a breaking change when it's necessary to fix a bug is IMHO not a problem and would be a good rationale to grant an SRU exception. I'll clarify this in the `What qualifies as SRU`

> ### Stable Release Update Procedure

Author: ```suggestion
### Point Release Procedure
```

At the moment, only a single major release will be supported, so all point releases will be based
off of that release.

In order to alleviate the burden for a single person to have to cherry-pick and handle merge conflicts

> of all desired backporting PRs to a point release, we instead maintain a living SRU branch, where

Reviewer 3: can we keep backport instead of SRU? it took me a sec to figure out what it meant.

Author: Please see my comment below

all desired features and bug fixes are merged into as separate PRs.

Example:

> Current release is `v0.38.4`. We then maintain a (living) branch `sru/release/v0.38.N`, given N as

Reviewer 3: we should try to avoid saying what the current release. We will have to amend this every time we do a release.

Author: ```suggestion
Given current release is `vMAJOR.MINOR.PATCH`. We then maintain a (living) branch `sru/release/vMAJOR.MINOR.PATCH+1`, given `N+1` as
```

Author: Does that work for you?

Author: @marbar3778 bump :)

Reviewer 3: yea that works. thanks

the next patch release number (currently `0.38.5`) for the `0.38` release series. As bugs are fixed
and PRs are merged into `master`, if a contributor wishes the PR to be released as SRU into the
`v0.38.N` point release, the contributor must:

> 1. Add `0.38-sru` label

Reviewer 3: i would prefer to keep `backport` this is another thing that we will have to update on every major release. This increases the maintenance load and cognitive load of everything that needs to be updated in these scenarios

Author: Alright, so instead of `backport` I'd propose to use a semantic versioning-like nomenclature: `MAJOR.MINOR.PATCH`

Author: The thing is, regardless of the name (we agreed that we stick with `backport`), I think that such living branch should carry the next version number to 1. avoid versioning mistakes 2. make clear to users what changes are going to be in exactly which release. Wdyt? @marbar3778 @alexanderbez

Reviewer 3: yes agree with you. `0.38-backport` works for me

Author: I actually meant "it should include the patch number" :) e.g. `0.38.5-backport`

Reviewer: The proposal of `0.38.5-backport` looks fine to me. I am infavor of calling it backport and not changing historical nomenclature.

@alessio  Can you update this file accordingly?

Author: Fixed. Please check

> 2. Pull latest changes on the desired `sru/release/v0.38.N` branch

Reviewer 3: ```suggestion
2. Pull latest changes on the desired `sru/release/vX.X.N` branch
```

off of that release.

In order to alleviate the burden for a single person to have to cherry-pick and handle merge conflicts

> of all desired backporting PRs to a point release, we instead maintain a living SRU branch, where

Author: ```suggestion
of all desired backporting PRs to a point release, we instead maintain a living backport branch, where
```

Example:

Current release is `v0.38.4`. We then maintain a (living) branch `sru/release/v0.38.N`, given N as
the next patch release number (currently `0.38.5`) for the `0.38` release series. As bugs are fixed
and PRs are merged into `master`, if a contributor wishes the PR to be released as SRU into the
`v0.38.N` point release, the contributor must:

> 1. Add `0.38-sru` label

Author: ```suggestion
1. Add `0.38.N-backport` label
```

Note, although we aim to support only a single release at a time, the process stated above could be
used for multiple previous versions.

> ## Stable Release Updates (SRU)

Author: ```suggestion
## Stable Release Updates
```

Author: Do you want me to rename this as well @alexanderbez @marbar3778? All in all, I don't think this naming is so bad, this should help clarify what backports are: updates to stable releases (aka maintenance releases). Wdyt?

Reviewer 2: I think we should stick with point releases.

Author: Done, thanks

Once a Cosmos-SDK release has been completed and published, updates for it are released under certain circumstances
and must follow the *Stable Release Update* (or SRU) procedure.

> ### Rationale

Author: @marbar3778 @alexanderbez here is the rationale. Thoughts? Does that make sense to you?

Reviewer 3: thank you for the rationale. I do agree but still dont understand the naming change. Every release is a stable release or should be. We can keep the rationale because it will help but there isnt a need for a rename IMO.

Author: What do you mean by every release? A `major.minor` is a major stable release (e.g. 0.38), backports/bugfixes/updates go into _point_ releases. Am I wrong?
